packages/eslint-plugin/tests/rules/no-magic-numbers.test.ts

```diff
@@ -58,6 +58,42 @@ class Foo {
       `,
       options: [{ ignoreReadonlyClassProperties: true }],
     },
+    {
+      code: 'type Foo = Bar[0];',
+      options: [{ ignoreTypeIndexes: true }],
+    },
+    {
+      code: 'type Foo = Bar[-1];',
+      options: [{ ignoreTypeIndexes: true }],
+    },
+    {
+      code: 'type Foo = Bar[0xab];',
+      options: [{ ignoreTypeIndexes: true }],
+    },
+    {
+      code: 'type Foo = Bar[5.6e1];',
+      options: [{ ignoreTypeIndexes: true }],
+    },
+    {
+      code: 'type Foo = Bar[10n];',
+      options: [{ ignoreTypeIndexes: true }],
+    },
+    {
+      code: 'type Foo = Bar[1 | -2];',
+      options: [{ ignoreTypeIndexes: true }],
+    },
+    {
+      code: 'type Foo = Parameters<Bar>[2];',
+      options: [{ ignoreTypeIndexes: true }],
+    },
+    {
+      code: "type Foo = Bar['baz'];",
+      options: [{ ignoreTypeIndexes: true }],
+    },
+    {
+      code: "type Foo = Bar['baz'];",
+      options: [{ ignoreTypeIndexes: false }],
+    },
```

Comment:
A few edge cases we'll also want to test for:

```ts
type Foo = Bar[1 & number];
```

```ts
type Other = {
    [0]: 3;
}

type Foo = {
  [K in keyof Other]: `${K & number}`;
};
```

```ts
type Foo = {
  [K in 0 | 1 | 2]: 0;
};
```

```ts
type Others = [ ['a'], ['b'] ];

type Foo = {
  [K in keyof Others[0]]: Others[K];
};
```

Reply:
@JoshuaKGoldberg couple of quick questions on those:

> ```ts
> type Foo = Bar[1 & number];
> ```

Sure, I check for union types so intersection should be covered too! Thanks.

> ```ts
> type Other = {
>     [0]: 3;
> }
> 
> type Foo = {
>   [K in keyof Other]: `${K & number}`;
> };
> ```

Since the `0` is a `TSPropertySignature`, not a `TSIndexedAccessType`, would you still want it to be covered by this new `ignoreTypeIndexes` config? Seems like that should be a separate config?

Also, the `3` there _should_ be an error, right? So perhaps `string` or some other non-numeric literal would be more appropriate for this test case?

> ```ts
> type Foo = {
>   [K in 0 | 1 | 2]: 0;
> };
> ```

As above, `TSPropertySignature`.

> ```ts
> type Others = [ ['a'], ['b'] ];
> 
> type Foo = {
>   [K in keyof Others[0]]: Others[K];
> };
> ```

I’ll add a test to ensure the `Others[0]` part is covered.

Reply:
Great questions!

> check for union types so intersection should be covered too

`Bar[1 & number]` is a bit different syntax from `${K & number}`. I think it'd be good to have an explicit test for both. The code right now might not differentiate, but it's ripe for future refactors to miss it.

> Since the 0 is a `TSPropertySignature`, not a `TSIndexedAccessType`, would you still want it to be covered by this new `ignoreTypeIndexes` config? 

I'm not requesting changes to logic, just that it be covered. Agreed that they're different things -- the tests should validate that `ignoreTypeIndexes` _doesn't_ remove the complaint on it.

> Also, the 3 there _should_ be an error, right?

Yes. That's already covered though; I'm really more interested in the `TSPropertySignature`.

Reply:
@JoshuaKGoldberg Thank you for clarifying. In af1ddb7a3cb042ed13b388278dc06fa1290b0afe I added support for intersection types, nesting of union/intersection types, and the additional invalid test cases you suggested.